queries/period_slippage.sql

```diff
@@ -310,70 +310,114 @@ incoming_and_outgoing_with_buffer_trades as (
-    from final_token_balance_sheet
-    left join prices.usd
-        on contract_address = token
-        and minute = '{{EndTime}}'
+        date_trunc('hour', minute) as hour,
+        minute,
+        token
+    from
+        final_token_balance_sheet
+    group by
+        hour,
+        minute,
+        token
 ),
--- The approximate prices are based on the hourly median prices over the past day.
-approximate_prices as (
+token_times_hour as (
     select
-        contract_address,
-        decimals,
-        percentile_cont(0.5) WITHIN group (order by median_price) as price
-    from prices.prices_from_dex_data p
-    where hour > '{{EndTime}}'::timestamptz - interval '1 day'
-    and date(hour) = '{{EndTime}}'::timestamptz - interval '1 day'
-      -- Getting approximate prices only for tokens we do not have an exact price for.
-    and contract_address IN (select contract_address from exact_prices where price is null)
-    group by contract_address, decimals
-    having sum(sample_size) > 0
+        hour,
+        token
+    from
+        token_times
+    group by
+        hour,
+        token
 ),
-end_prices as (
-    select * from exact_prices where price is not null
-    union
-    select * from approximate_prices
+precise_prices as (
+    select
+        pusd.*
+    from
+        prices.usd pusd
+        inner join token_times tt on pusd.minute = tt.minute
+        and contract_address = token
+),
+median_prices as (
+    select
+        musd.*
+    from
+        prices.prices_from_dex_data musd
+        inner join token_times_hour tt on musd.hour = tt.hour
+        and contract_address = token
+),
+prices as (
+    select
+        minute,
+        COALESCE(
+            precise.contract_address,
+            median.contract_address
+        ) as contract_address,
+        COALESCE(precise.decimals, median.decimals) as decimals,
+        case
+            when price is not null then price
+            else median_price
+        end as price
+    from
+        precise_prices precise
+        left join median_prices median on date_trunc('hour', minute) = hour
+        and precise.contract_address = median.contract_address
 ),
 results_per_tx as (
-    select solver_address,
-           solver_name,
-           sum(token_imbalance_wei * price / 10 ^ p.decimals) as usd_value,
-           tx_hash
-    from final_token_balance_sheet
-             inner join end_prices p on token = p.contract_address
-    group by solver_address,
-             solver_name,
-             tx_hash
-    having sum(token_imbalance_wei) != 0
+    select
+        solver_address,
+        solver_name,
+        sum(token_imbalance_wei * price / 10 ^ p.decimals) as usd_value,
+        tx_hash
+    from
+        final_token_balance_sheet ftbs
+        left join prices p on token = p.contract_address
+        and p.minute = ftbs.minute
+    group by
+        solver_address,
+        solver_name,
+        tx_hash
+    having
+        bool_and(price is not null)
```

Comment:
Are wet worried that Solvers could start filling batches with trades containing unknown token prices and this always be excluded from penalty?

Reply:
In the example you shared we are omitting a valid penalty. Are there example transactions where we would penalize incorrectly?

Reply:
But is the punishment not too strong? I think the solver did something illegal, but punishing it by the whole 380 dollars might be too tough?
We never defined a real punishment and if we do a punishment, then we would also have to take it from the bonding pool?

Reply:
> But is the punishment not too strong?

I don't think so. The damage is exactly the current value of the tile (assuming it could go to 0). I also don't think we need to take any penalty from the bonding pools as long as we can withhold rewards (first ETH, then COW). We can allow the solver to withdraw what is left from that token but I'm not sure we should do that for them. We really don't want to encourage this behaviour so I'd err on the side of slightly higher penalty.

Reply:
Given that soon all solvers will be reporting which interactions are replaced with buffers (which means that we could simply reject all solutions that do illegal buffer trading, like PLM did this week), I guess the main reason for non-safe tokens to end up to our contract would be slippage.

And in this case, one option is to go (as we did up till today) with the "price" of that token at the day of the payout. Of course, the main challenge in these examples is to obtain some source of truth for what the price of each such token is. In any case, if we keep this strategy, then we could just as well apply it here for now, and maybe we can just add some additional eth as penalty, in order to cover our cost of selling the token (in case it makes sense for us to sell it). For the case discussed above, I agree with Alex that charging the whole amount might be a bit too much, at least for the cases the token we currently keep still has some value.

An extreme solution would be to maintain some accounting for each solver for "next week". We would then first sell all the random tokens, see the actual executed price we got, add gas costs, and all this would be charged to the solver the week after.

Reply:
>Are we worried that Solvers could start filling batches with trades containing unknown token prices and this always be excluded from penalty?

I think if prices are missing, then can theoretically trick us the on or the other way. If we omit the batches, then they can do negative slippage. If we do incorrect accounting, then they can trade such that they benefit from the incorrect accounting. I would prefer to stick to the omitting strategy and then later punish the solvers, if we see that they abused it

Reply:
>re there example transactions where we would penalize incorrectly?

I could not find any example